src/navigation/navigationstate.cpp

```diff
@@ -0,0 +1,186 @@
+ * permit persons to whom the Software is furnished to do so, subject to the following   *
+ * conditions:                                                                           *
+ *                                                                                       *
+ * The above copyright notice and this permission notice shall be included in all copies *
+ * or substantial portions of the Software.                                              *
+ *                                                                                       *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,   *
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A         *
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT    *
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF  *
+ * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE  *
+ * OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                                         *
+ ****************************************************************************************/
+
+#include <openspace/camera/camerapose.h>
+#include <ghoul/logging/logmanager.h>
+#include <openspace/navigation/navigationstate.h>
+#include <openspace/scene/scenegraphnode.h>
+#include <openspace/query/query.h>
+
+namespace {
+    constexpr const char* _loggerCat = "NavigationState";
+
+    const double Epsilon = 1E-7;
+
+    struct [[codegen::Dictionary(NavigationState)]] Parameters {
+        // The identifier of the anchor node
+        std::string anchor;
+
+        // The identifier of the aim node, if used
+        std::optional<std::string> aim;
+
+        // The identifier of the scene graph node to use as reference frame. If not
+        // specified, this will be the same as the anchor
+        std::optional<std::string> referenceFrame;
+
+        // The position of the camera relative to the anchor node, expressed in meters in
+        // the specified reference frame
+        glm::dvec3 position;
+
+        // The up vector expressed in the coordinate system of the reference frame
+        std::optional<glm::dvec3> up;
+
+        // The yaw angle in radians. Positive angle means yawing camera to the right
+        std::optional<double> yaw;
+
+        // The pitch angle in radians. Positive angle means pitching camera upwards
+        std::optional<double> pitch;
+    };
+#include "navigationstate_codegen.cpp"
+} // namespace
+
+namespace openspace::interaction {
+
+NavigationState::NavigationState(const ghoul::Dictionary& dictionary) {
+    const Parameters p = codegen::bake<Parameters>(dictionary);
+
+    anchor = p.anchor;
+    position = p.position;
+
+    referenceFrame = p.referenceFrame.value_or(anchor);
+    aim = p.aim.value_or(aim);
+
+    if (p.up.has_value()) {
+        up = *p.up;
+
+        yaw = p.yaw.value_or(yaw);
```

Comment:
Should these really be inside the if statement? What if `yaw` and `pitch` is provided but not `up`? And if `up` is required in order to use `yaw` and `pitch` then it should be beter documented to avoid confusion.

Reply:
I agree it's not the best design, but this is how it was done before and none of the three values makes sense without the others in the case of a NavigationState. 

A better option is to combine the three angles into an EulerAngles struct in the Parameters. That would make it easy to handle it correctly in terms of testing against the specification as well (right now, just adding one of the three would not result in an error message). However, that would be a breaking change that could potentially affect quite a few users

As it's not a change made in this code, but rather just a function that has been copied, I suggest keeping it as it is for this PR. If we want to change it, we can create an issue for it. Then the change would be properly "documented" within our GitHub history